crates/worker/src/partition_processor_manager.rs

```diff
@@ -8,94 +8,170 @@
 use restate_partition_store::{OpenMode, PartitionStore, PartitionStoreManager};
 use restate_types::arc_util::ArcSwapExt;
 use restate_types::config::{UpdateableConfiguration, WorkerOptions};
 use restate_types::epoch::EpochMetadata;
 use restate_types::identifiers::{LeaderEpoch, PartitionId, PartitionKey};
 use restate_types::logs::{LogId, Payload};
 use restate_types::metadata_store::keys::partition_processor_epoch_key;
-use restate_types::{GenerationalNodeId, Version};
+use restate_types::time::MillisSinceEpoch;
+use restate_types::GenerationalNodeId;
 use restate_wal_protocol::control::AnnounceLeader;
 use restate_wal_protocol::{Command as WalCommand, Destination, Envelope, Header, Source};
-use std::collections::HashMap;
-use std::ops::RangeInclusive;
-use tokio::sync::mpsc;
-use tracing::{debug, info};
+
+use crate::partition::storage::invoker::InvokerStorageReader;
+use crate::partition::PartitionProcessorControlCommand;
+use crate::PartitionProcessor;
 
 pub struct PartitionProcessorManager {
     updateable_config: UpdateableConfiguration,
-    running_partition_processors: HashMap<PartitionId, TaskId>,
+    running_partition_processors: BTreeMap<PartitionId, State>,
 
     metadata: Metadata,
     metadata_store_client: MetadataStoreClient,
     partition_store_manager: PartitionStoreManager,
+    attach_router: RpcRouter<AttachRequest>,
     networking: Networking,
     bifrost: Bifrost,
     invoker_handle: InvokerHandle<InvokerStorageReader<PartitionStore>>,
     rx: mpsc::Receiver<ProcessorsManagerCommand>,
     tx: mpsc::Sender<ProcessorsManagerCommand>,
+    latest_attach_response: Option<(GenerationalNodeId, AttachResponse)>,
+}
+
+#[derive(Debug, thiserror::Error)]
+enum AttachError {
+    #[error("No cluster controller found in nodes configuration")]
+    NoClusterController,
+    #[error(transparent)]
+    ShutdownError(#[from] ShutdownError),
+}
+
+struct State {
+    _created_at: MillisSinceEpoch,
+    _key_range: RangeInclusive<PartitionKey>,
+    _control_tx: mpsc::Sender<PartitionProcessorControlCommand>,
+    _watch_rx: watch::Receiver<PartitionProcessorStatus>,
+    _task_id: TaskId,
 }
 
 impl PartitionProcessorManager {
+    #[allow(clippy::too_many_arguments)]
     pub fn new(
         updateable_config: UpdateableConfiguration,
         metadata: Metadata,
         metadata_store_client: MetadataStoreClient,
         partition_store_manager: PartitionStoreManager,
+        router_builder: &mut MessageRouterBuilder,
         networking: Networking,
         bifrost: Bifrost,
         invoker_handle: InvokerHandle<InvokerStorageReader<PartitionStore>>,
     ) -> Self {
+        let attach_router = RpcRouter::new(networking.clone(), router_builder);
+
         let (tx, rx) = mpsc::channel(updateable_config.load().worker.internal_queue_length());
         Self {
             updateable_config,
-            running_partition_processors: HashMap::default(),
+            running_partition_processors: BTreeMap::default(),
```

Comment:
Is `BTreeMap` faster than `HashMap` for small number of entries?

Reply:
Correct, it's also more cacheline friendly, and in general better when keys are small and numeric.